File: blas/hint32_arm64.go
package blas

import (
	"unsafe"
)

//go:noescape

func _hsum_int32(input, result unsafe.Pointer, size uint64)
func HsumInt32(input []int32) int32 {
	var sum int32
	_hsum_int32(unsafe.Pointer(&input[0]), unsafe.Pointer(&sum), uint64(len(input)))
	return sum
}

func _hmax_int32(input, result unsafe.Pointer, size uint64)
func HmaxInt32(input []int32) int32 {
	var max int32
	_hmax_int32(unsafe.Pointer(&input[0]), unsafe.Pointer(&max), uint64(len(input)))
	return max
}

func _hmin_int32(input, result unsafe.Pointer, size uint64)
func HminInt32(input []int32) int32 {
	var min int32
	_hmin_int32(unsafe.Pointer(&input[0]), unsafe.Pointer(&min), uint64(len(input)))
	return min
}

File: blas/cpp/hint32_arm64.cpp
#include <arm_neon.h>

void hsum_int32(int32_t *input, int32_t *result, uint64_t size) {
    *result = 0.0;
    int chunks = size / 4;

    for (int i = 0; i < chunks; i++) {
        *result += vaddvq_s32(((int32x4_t *) input)[i]);
    }
    // Leftovers
    for (int i = chunks * 4; i < size; i++) {
        *result += input[i];
    }
}

void hmax_int32(int32_t *input, int32_t *result, uint64_t size) {
    *result = input[0];
    int chunks = size / 4;

    for (int i = 0; i < chunks; i++) {
        int32_t tmp = vmaxvq_s32(((int32x4_t *) input)[i]);
        if (tmp > *result) {
            *result = tmp;
        }
    }
    // Leftovers
    for (int i = chunks * 4; i < size; i++) {
        if (input[i] > *result) {
            *result = input[i];
        }
    }
}

void hmin_int32(int32_t *input, int32_t *result, uint64_t size) {
    *result = input[0];
    int chunks = size / 4;

    for (int i = 0; i < chunks; i++) {
        int32_t tmp = vminvq_s32(((int32x4_t *) input)[i]);
        if (tmp < *result) {
            *result = tmp;
        }
    }
    // Leftovers
    for (int i = chunks * 4; i < size; i++) {
        if (input[i] < *result) {
            *result = input[i];
        }
    }
}
